src/densities.jl

```diff
@@ -138,13 +138,12 @@ total_density(ρ) = dropdims(sum(ρ; dims=4); dims=4)
 end
 
 function ρ_from_total_and_spin(ρtot, ρspin=nothing)
-    n_spin = ρspin === nothing ? 1 : 2
-    ρ = similar(ρtot, size(ρtot)..., n_spin)
-    if n_spin == 1
-        ρ .= ρtot
+    if ρspin === nothing
+         # returns a copy for consistency with the other case
+        copy(reshape(ρtot, (size(ρtot)..., 1)))
     else
-        ρ[:, :, :, 1] .= (ρtot .+ ρspin) ./ 2
-        ρ[:, :, :, 2] .= (ρtot .- ρspin) ./ 2
+        # Val to ensure inferability
```

Comment:
Can't one ensure that by providing the return type? I mean if we capture the element type of `ρtot` with a `where` than the return type is sort of clear or not?

Reply:
Not sure if that's easier to parse though. Probably Val is still the better option.

Reply:
not sure that's a good idea, because eg if rho is a GPU array it could be that cat still lives on the GPU so we don't want to assert it's an Array

Reply:
yes cat still lives on the GPU I checked that.